### ISSUE TITLE
SABnzbd rating support

### ISSUE THREAD

**Issue by Safihre** [nZEDb/nZEDb#2235]
I'm from the SABnzbd team and I was wondering if you were aware that SABnzbd has build-in automatic indexer feedback system?

The automated feedback has been there since version 0.7.0 (about 4 years or so), allowing to send automatically back to the indexer if the download was successful and if not, from which servers it is missing or if it was encrypted. It also allows users to quickly rate the audio/video quality of a release from within SABnzbd. 
It was named OZnzb in our interface for years, since they started it, but for the upcoming release of SABnzbd (v1.1.0) **all references to OZnzb will be removed** and users can easily change in the settings what indexer-host they use.

I made a page on our wiki describing the workings of the system:
https://sabnzbd.org/wiki/extra/indexer-feedback 

Would you be interested in integrating this?
I understand that it might be not very useful for people running it just for local indexer, this would only be interesting when there's a lot of users. Just wanted to let you guys know what SAB can do! 
(and appreciate all the work you guys do to keep those indexers going!)

**Comment by DariusIII** [comment]
As we discussed already on reddit, you need to at least adapt that path to page sab expects. If you expect us to support it, you need to be a bit more flexible and willing to adapt.
This is my opinion, other devs might have a different one.

**Comment by niel** [comment]
Yes we were aware of the feature. However until recently there was no documentation on it. I briefly read the documentation and have to take issue with one statement:

> SABnzbd has automatic feedback for indexers integrated since before version 0.7. It was originally made for OZnzb and thus had these labels within SABnzbd. However, it has always supported other indexers and since SABnzbd 1.1.0 the references to OZnzb have been removed.

It only supports indexers that follow OZnzb's model. We do not.

We have a published API (which complies with Newznab's) that we would be happy to add a new entry to, for this feature. However we don't have the time, or desire, to create an exception that may prove to be a maintenance headache in the future if/when we move away from Smarty as our template engine.

If/when we get to a point where adding an entry into a routing table would make this simple, we will reconsider.

Again, this is my opinion other devs may have a different one.

BTW, we do consider all Pull Requests for new features provided they comply with our Coding Standards.

**Comment by Safihre** [comment]
So what would make it easier? If the API endpoint URL called by Sabnzbd would be different? 
It's completely independent from OZNZB, and the code has remained the same for 4 years so changes are very unlikely, if this is a concern. 

It will require information inside the NZB so Sabnzbd  knows it's a job that supports rating,  not just an API endpoint (see the wiki page).

**Comment by kevinlekiller** [comment]
Where does the rating-id come from ? `<meta type="x-rating-id">ID-OF-JOB</meta>`

Can it be just some unique string associated to that nzb for the indexer?

Also `https://RATING_HOST/releaseRatings/releaseRatings.php`

Does it need to end in .php?

**Comment by DariusIII** [comment]
I believe it is an id indexer provides (ie. guid of the release) so when sab gets back to indexer it provides same id to work with.

**Comment by kevinlekiller** [comment]
That's what I figured, but I think that should be documented.

**Comment by DariusIII** [comment]
I agree.

**Comment by kevinlekiller** [comment]
I have no objections to this being implemented fyi, I'm just wondering how the .php will work out with the URL rewriting.

It's also an issue if someone were writing an indexer in a language other than PHP.

**Comment by niel** [comment]
I get the impression they want us to add an exception to the rewrite rules for their URL and have a PHP file at the URL to process their request.
IMO too much hassle for a limited benefit. Most of our users run private indexes. This feature only benefits indexing sites with larger numbers of users. If that is so, the site operators can implement the feature if it matters to them.
Having said that, a good PR could change my opinion.

**Comment by Safihre** [comment]
We can definitely change the endpoint, it ending on .php is indeed limiting and not very open. I can make some fallback in the code. 

What would you suggest? 
`/rating` (just a suggestion) 

The question to me seemed indeed more about the value for your users, are enough of them public indexers that can really benefit from it? 
I just thought if it's already part of the base configuration, it lowers the threshold for public indexers to use it :)

**Comment by kevinlekiller** [comment]
Could the URL maybe be like it is currently (maybe a array of sane values - instead of a single string), but can also be set in the NZB, and/or through the HTTP headers (similar to the `X-DNZB-Failure`)?

Edit: Reason why I'm thinking of the headers is in case for example the NZB is already written (newznab/nzedb store the NZB on the disk), but the hostname changes.

Thanks.

**Comment by niel** [comment]
Nice idea. Have the target URL definable by the indexer in headers and SABnzbd defaults to currently used address if none specified. That might work for us.

**Comment by Safihre** [comment]
How about putting it in the meta data of the NZB? 
 This way the information is kept even when the NZB is downloaded first, just like the x-rating-id. 
There is also a `x-rating-host` meta tag (not described yet on the wiki page because it didn't seem too usefull to me), which allows to override the host set in the Config.

**Comment by niel** [comment]
Better in the header, I think. The information is transitory, i.e. the URL to reply to might change. It is even conceivable we might want to generate unique addresses to reply to for different requests. Storing a response-url in the NZB would mean deciding long before the file is ever requested what URL it should respond on. It also assumes the same URL would be used for all responses for that NZB, with no possibility of spreading them across multiple front-ends for example.
If SAB wants to modify the NZB to store the data in there it can, but it should realise that the address has a limited usability and the server may refuse the connection  after some arbitrary period.

**Comment by Safihre** [comment]
Seems reasonable and more future proof.
I will add the analysis of the header to accept all the fields that are also available in the meta-tag plus the `x-rating-host` and `x-rating-url` (where the URL can be a full URL, starting with `https`, or relative to the `x-rating-host`). Will update the Wiki afterwards.

So these fields in header or meta, whatever the indexer likes:

``` html
<meta type="x-rating-id">ID-OF-JOB-FROM-INDEXER</meta>
<meta type="x-rating-url">URL-TO-REPORT-TO</meta>
<meta type="x-rating-host">HOST-OF-INDEXER</meta>
<meta type="x-rating-video">VIDEO-SCORE</meta> <!-- From 1-10 -->
<meta type="x-rating-videocnt">NUMBER-OF-VOTES</meta>
<meta type="x-rating-audio">AUDIO-SCORE</meta>
<meta type="x-rating-audiocnt">NUMBER-OF-VOTES</meta>
<meta type="x-rating-voteup">NUMBER-OF-UPVOTES</meta>
<meta type="x-rating-votedown">NUMBER-OF-DOWNVOTES</meta>
<meta type="x-rating-spam">NUMBER-OF-TIMES-REPORTED-AS-SPAM</meta> 
<meta type="x-rating-confirmed-spam">yes/no</meta>  <!-- Confirmed by admins/mods -->
<meta type="x-rating-passworded">NUMBER-OF-TIMES-REPORTED-AS-PASSWORDED</meta> 
<meta type="x-rating-confirmed-passworded">yes/no</meta>  <!-- Confirmed by admins/mods -->
```

**Comment by niel** [comment]
Looks good. Only thing I would consider adding is x-rating-url-ttl for indexers that want to specifiy how long the url is valid for (defaulting to 0, meaning no specified limit). It should be in minutes I guess, although having a value that gives SAB a reasonable chance to download the release and be able to respond.

**Comment by DariusIII** [comment]
I would suggest that responses from SABNzbd should be aggregated into one array. Something like:

```
{  
    "apikey": "KEY-SET-IN-CONFIG",
    "i": "ID-OF-JOB",
    "sa": "NEWSSERVER-ADDRESS",
    "m" : "1 or 0",
    "auto": "1 OR 0",
    "rp": "1 or 0",
    "o" : "text",
    "rc" : "user text",
    "vq" : "1 to 10",
    "aq" : "1 to 10"
}
```

Or similar to this. The way it is currently, if i read the docs properly,  requires sab to contact indexer for every rating, comment etc separately.

**Comment by Safihre** [comment]
Yes, because these are separate user actions and are not send as a batch.

EDIT:
For example a user can upvote it to let the indexer know the download was fine, and after watching it (hours/days later) give a specific audio/video rating or a comment.

**Comment by DariusIII** [comment]
I suggest you make it into one call, per release downloaded, to make it less stressful for indexers (making 4, 5 or more calls for just one download is bad). I believe other devs will agree with me.

**Comment by Safihre** [comment]
You are now asking us to change the whole API and not just extend it. 
The API is based on the idea that rating actions don't happen all at once and the action (`mode`) is therefore specified in the `m` parameter. So the `r` (the rating) parameter can be a comment, a problem or a rating, based on the `mode`.
Realistically, you will get 2, maybe 3, API calls per download if a user decides to rate and comment.

**Comment by DariusIII** [comment]
You have 8 possible responses  from sabnzbd to indexer, imagine that at least 4 are made for same release? It is 4 calls from one user. I will take an example of my own testing indexer i opened up for public recently to have better "influx" on development of newznab-tmux and nZEDb. I have 1,5k users atm, let's say 500 will use the feature. That is 2000 possible hits if they decide to rate 1 downloaded release each, and it could be just 500. And that is just for a small indexer. There are ones with 20-50k users. I believe they will rather not use the feature than have their servers brought down to their knees because of it.

What i want to say is, response from sabnzbd needs to be optimized. Of course, this is my personal opinion and other devs might not agree with it.

**Comment by kevinlekiller** [comment]
Even if it was all grouped it wouldn't change the amount of requests, since the person who downloaded might not vote for hours or days after it is downloaded, since most people do not review the content as soon as it's done downloading.

By grouping it all up it would increase the amount of data sent to the indexer.

**Comment by kevinlekiller** [comment]
> What i want to say is, response from sabnzbd needs to be optimized. 

I guess what you mean to say is you want it to buffer, for example the person presses vote up, then you want Sabnzbd to wait x seconds in case the person click spam before sending the request (with both vote + spam in a single request)?

**Comment by DariusIII** [comment]
User can do it how ever he pleases, but sab should send it as one response, at least ones that can be grouped, like video and audio quality and votes.

**Comment by DariusIII** [comment]
@Safihre 
As examples on sabnzbd page are not enough for me, i would like to see an complete rating sent to indexer by sabnzbd. If we put my previous requset aside for the moment, i would like to have more informative output from sab. Having m: pr and then pr: something does not help. Can the m part be removed? 

ie:

```
{
    "m": "o", 
    "r": "USER-TEXT"
}
```

would become:

```
{
    "o": "USER-TEXT"
}
```

and one more:

```
{  
    "m": "r",
    "r": "videoQuality",
    "rn": "RATING 1-10"
}
```

would become:

```
{  
    "vq": "RATING 1-10"
}
```

Is this doable?

**Comment by Safihre** [comment]
Regarding the testing, we build a new Release Candidate (RC4) yesterday that has the changes as proposed before today. Maybe this will make it easier :)
https://sabnzbd.org/downloads

**Comment by DariusIII** [comment]
What about the second part of my question?

**Comment by Safihre** [comment]
Have to talk/think about that since it's a complete change of this API that has been around for years, but hidden.
The intention for SABnzbd 1.1.0 was to extend the current API, not completely overhaul it. Especially since we also send this information to big indexers and shared it on Reddit etc.

**Comment by niel** [comment]
I took a new look at your "Indexer feedback" wiki page today. 
While it does contain examples, it really can not be called API documentation.
For example, you use 

> {
>     "m": "rp",
>     "auto": "1 OR 0"
> }

as an example, but no where do you explain what "m" means. In other words you are not defining the API at all, just showing examples of particular ways to use it. This makes designing an appropriate way to handle it difficult at best.

I have to agree with @DariusIII that the "m": "<method>", "<option_name>": "<option_value>" syntax is clunky/burdonsome and an array of methods/options would be far more useful. Although I don't believe it will have much effect on requests, it would definitely make implementing an API easier.

Currently I am not interested in seeing this implemented because of these issues. However, @DariusIII is putting the effort into it, and I wouldn't dream of preventing him from doing so.
